subprojects/base-services/src/main/java/org/gradle/api/InvalidUserDataException.java

```diff
@@ -16,9 +16,12 @@
 
 package org.gradle.api;
 
+import org.gradle.internal.exceptions.Contextual;
+
 /**
  * A <code>InvalidUserDataException</code> is thrown, if a user is providing illegal data for the build.
  */
+@Contextual
```

Comment:
I would like to understand better the consequence of this change.
Can you share your investigation?

Reply:
So my investigation was like this. I saw that when the failure happens an exception is thrown, specifically the `GradleException` from below (which will end up being wrapped by various other exceptions, but we can ignore that for now), which has as its cause an `InvalidUserDataException`, which in turn has as its cause another `InvalidUserDataException` and these two `InvalidUserDataException`'s messages describe the root cause of the failure, but these messages don't get logged by Gradle.

When I investigated why that is, I ended up in `ContextAwareException.visitCause()`, then in the `findNearestContextual()` method, where it becomes obvious. 

That's how I concluded that I would need to annotate `InvalidUserDataException` with the `Contextual` annotation, to make their message part of the cause logged by Gradle.

Reply:
I think annotating `InvalidUserDataException` with contextual will change quite a few exception messages, I am not sure that is good or bad. It seems like `InvalidUserDataException` should not add context to an already existing exception, that is why it is probably not marked as `@Contextual`. Instead, it should provide the full message why something is invalid user data. Maybe it is used wrongly for the dependency verification exception?

Reply:
I think that if replacing the `InvalidUserDataException` with a dedicated dependency verification exception which is `@Contextual` resolves the issue, it feels like a better change.

Reply:
Will look into such an approach then.

Reply:
Looks much better the way Louis suggested and feels much safer indeed.